Source/Core/Core/PowerPC/PPCCache.cpp

```diff
@@ -94,169 +94,338 @@ InstructionCache::~InstructionCache()
+  auto& memory = system.GetMemory();
+
+  auto* value = static_cast<const u8*>(buffer);
+
+  while (len > 0)
+  {
+    auto [set, way] = GetCache(addr, locked);
+
+    u32 offset_in_block = addr - (addr & ~31);
+    u32 len_in_block = std::min<u32>(len, ((addr + 32) & ~31) - addr);
+
+    if (way != 0xff)
+    {
+      std::memcpy(reinterpret_cast<u8*>(data[set][way].data()) + offset_in_block, value,
+                  len_in_block);
+      modified[set] |= (1 << way);
+    }
+    else
+    {
+      memory.CopyToEmu(addr, value, len_in_block);
+    }
+
+    addr += len_in_block;
+    len -= len_in_block;
+    value += len_in_block;
+  }
+}
+
+void Cache::DoState(PointerWrap& p)
 {
   if (p.IsReadMode())
   {
     // Clear valid parts of the lookup tables (this is done instead of using fill(0xff) to avoid
     // loading the entire 4MB of tables into cache)
-    for (u32 set = 0; set < ICACHE_SETS; set++)
+    for (u32 set = 0; set < CACHE_SETS; set++)
     {
-      for (u32 way = 0; way < ICACHE_WAYS; way++)
+      for (u32 way = 0; way < CACHE_WAYS; way++)
       {
         if ((valid[set] & (1 << way)) != 0)
         {
-          const u32 addr = (tags[set][way] << 12) | (set << 5);
-          if (addr & ICACHE_VMEM_BIT)
-            lookup_table_vmem[(addr >> 5) & 0xfffff] = 0xff;
-          else if (addr & ICACHE_EXRAM_BIT)
-            lookup_table_ex[(addr >> 5) & 0x1fffff] = 0xff;
+          if (addrs[set][way] & CACHE_VMEM_BIT)
+            lookup_table_vmem[((addrs[set][way] >> 5) & 0xfff80) | set] = 0xff;
+          else if (addrs[set][way] & CACHE_EXRAM_BIT)
+            lookup_table_ex[((addrs[set][way] >> 5) & 0x1fff80) | set] = 0xff;
           else
-            lookup_table[(addr >> 5) & 0xfffff] = 0xff;
+            lookup_table[((addrs[set][way] >> 5) & 0xfff80) | set] = 0xff;
         }
       }
     }
   }
 
   p.DoArray(data);
-  p.DoArray(tags);
   p.DoArray(plru);
   p.DoArray(valid);
+  p.DoArray(addrs);
+  p.DoArray(modified);
 
   if (p.IsReadMode())
   {
     // Recompute lookup tables
-    for (u32 set = 0; set < ICACHE_SETS; set++)
+    for (u32 set = 0; set < CACHE_SETS; set++)
     {
-      for (u32 way = 0; way < ICACHE_WAYS; way++)
+      for (u32 way = 0; way < CACHE_WAYS; way++)
       {
         if ((valid[set] & (1 << way)) != 0)
         {
-          const u32 addr = (tags[set][way] << 12) | (set << 5);
-          if (addr & ICACHE_VMEM_BIT)
-            lookup_table_vmem[(addr >> 5) & 0xfffff] = way;
-          else if (addr & ICACHE_EXRAM_BIT)
-            lookup_table_ex[(addr >> 5) & 0x1fffff] = way;
+          if (addrs[set][way] & CACHE_VMEM_BIT)
+            lookup_table_vmem[((addrs[set][way] >> 5) & 0xfff80) | set] = 0xff;
+          else if (addrs[set][way] & CACHE_EXRAM_BIT)
+            lookup_table_ex[((addrs[set][way] >> 5) & 0x1fff80) | set] = 0xff;
           else
-            lookup_table[(addr >> 5) & 0xfffff] = way;
+            lookup_table[((addrs[set][way] >> 5) & 0xfff80) | set] = 0xff;
```

Comment:
That said, it doesn't seem like you're taking advantage of that property anywhere other than immediately before the `addrs[set][way] = addr` line. I *think* this can now be written as just:

```C++
          if (addrs[set][way] & CACHE_VMEM_BIT)
            lookup_table_vmem[(addrs[set][way] >> 5) & 0xfffff] = 0xff;
          else if (addrs[set][way] & CACHE_EXRAM_BIT)
            lookup_table_ex[(addrs[set][way] >> 5) & 0x1fffff] = 0xff;
          else
            lookup_table[(addrs[set][way] >> 5) & 0xfffff] = 0xff;
```

Reply:
Yeah that is better, I didn't realize that